core/src/main/java/org/apache/calcite/rel/rules/ReduceExpressionsRule.java

```diff
@@ -738,6 +740,20 @@ protected static boolean reduceExpressionsInternal(RelNode rel,
     // Replace predicates on CASE to CASE on predicates.
     boolean changed = new CaseShuttle().mutate(expList);
 
+    // distinct for IN rexNode
+    for (RexNode rexNode : predicates.pulledUpPredicates) {
+      if (rexNode.isA(SqlKind.IN)) {
+        RexSubQuery rexSubQuery = (RexSubQuery) rexNode;
+        List<ImmutableList<RexLiteral>> collect =
+            ((LogicalValues) rexSubQuery.rel).tuples
+                .stream().distinct().collect(Collectors.toList());
+        // In would transform to RexSubQuery and rel should be LogicalValues
+        if (rexSubQuery.rel instanceof LogicalValues) {
+          ((LogicalValues) rexSubQuery.rel).tuples = ImmutableList.copyOf(collect);
```

Comment:
you cannot mutate the fields of a subquery; that subquery may be used by other objects as well.
you have to create a new object.

Reply:
OK,  would change it as your suggesstion. @mihaibudiu

Reply:
@mihaibudiu According my analysis，there is a problem  create a new object of SubQuery instead of  mutating the fields of a subquery ,since pulledUpPredicates is final ImmutableList<RexNode>. Maybe we cannot  change the construct of predicate by this way.

Reply:
Just allocate a new instance of LogicalValues, or use the builder to build one